### PR TITLE
Use separate artifact name per step

### PR DESCRIPTION
**Pull request by srittau**
Compatibility with actions/upload-artifact@v4

Cf. #11684

### CODE REVIEW

**Comment by srittau** [comment]
The primer comment run fails: https://github.com/python/typeshed/actions/runs/8508947187/job/23303502602, probably since it's still using the "main" version of itself.

File: .github/workflows/mypy_primer_comment.yml [hunk 1]
```diff
@@ -27,18 +27,20 @@ jobs:
...
-            });
-            fs.writeFileSync("diff.zip", Buffer.from(download.data));
+            for (const matchArtifact of matchArtifacts) {
+              const download = await github.rest.actions.downloadArtifact({
+                 owner: context.repo.owner,
+                 repo: context.repo.repo,
+                 artifact_id: matchArtifact.id,
+                 archive_format: "zip",
+              });
+              fs.writeFileSync(`diff_${matchArtifact.name}.zip`, Buffer.from(download.data));
```
[the remark this answers is not in the document]
Comment (srittau) [reply .github/workflows/mypy_primer_comment.yml:40]: I probably does, but I just wanted a quick way to have unique filenames. We can just use `${matchArtifact.name}.zip`, I guess.

File: .github/workflows/mypy_primer_comment.yml [hunk 2]
```diff
@@ -27,18 +27,20 @@ jobs:
...
+                 owner: context.repo.owner,
+                 repo: context.repo.repo,
+                 artifact_id: matchArtifact.id,
+                 archive_format: "zip",
+              });
+              fs.writeFileSync(`${matchArtifact.name}.zip`, Buffer.from(download.data));
+            }
 
-      - run: unzip diff.zip
+      - run: for file in *.zip; do unzip $file; done
```
Comment (JelleZijlstra) [inline .github/workflows/mypy_primer_comment.yml:43]: Does this not work?
```suggestion
      - run: unzip *.zip
```

Comment (srittau) [reply .github/workflows/mypy_primer_comment.yml:43]: No. `unzip` expects the first argument to be a zip file and the remaining arguments to be archive members to unzip.

**Pull request merged**